### ISSUE TITLE
Confused about reduce_fx in self.log

### ISSUE THREAD

**Issue by yipliu** [ashleve/lightning-hydra-template#436]
Hi,

I was confused when I read about the way [reduce_fx](https://pytorch-lightning.readthedocs.io/en/stable/extensions/logging.html) is used.
> Reduction function over step values for the end of an epoch. Uses torch.mean() by default and is not applied when a [torchmetrics.Metric](https://torchmetrics.readthedocs.io/en/stable/pages/implement.html#torchmetrics.Metric) is logged.

Does this mean that the loss will be averaged, but the torchmetrics.Metric will remain the same?

**Comment by ashleve** [comment]
`reduce_fx` is the method of reduction when logging values **directly**:
```python
acc = self.train_acc.compute()
self.log("train/acc", acc, reduce_fx="mean")
``` 

But when you log values through **metric object**, the reduce_fx doesn't do anything as far as I'm aware:
```python
self.log("train/acc", self.train_acc)
```

This is because torchmetrics have their own internal strategy of reducing over the epoch.

**Comment by ashleve** [comment]
Yesterday I changed the way loss and accuracy are logged with this PR: https://github.com/ashleve/lightning-hydra-template/pull/426
(you can read about reasoning in the description)

To verify correctness, I've run the template with loss logging in two ways:
```python
self.log("val/loss_v1", loss, on_step=True, on_epoch=True)
self.log("val/loss_v2", self.val_loss, on_step=True, on_epoch=True)
```

And the result seems to be exactly the same in both cases:
![image](https://user-images.githubusercontent.com/43458894/189109605-d05576a4-dca9-4a21-8d7b-f6591dfdff79.png)

Is there something I'm missing here?

The code you attached:
```
self.value = self.value + value * batch_size
self.cumulated_batch_size = self.cumulated_batch_size + batch_size

real_record = self.value / self.cumulated_batch_size 
```
The `self.value / self.cumulated_batch_size` is the way metric is logged at the end of epoch when `on_epoch=True`. I believe this is the same for both logging through metric and logging through value.

But when `on_step=True`, only the latest value is recorded (not the mean across all latest steps).

**Comment by ashleve** [comment]
It's a reduction over epoch strategy which applies only when passing **value** directly to `self.log(...).

If you instead pass metric object:
- the reduction will happen according to given metric internal strategy
- the metric will get automatically resetted by lightning when epoch ends
